### DISCUSSION TITLE
Calling `revalidateTag` / `revalidatePath` in a Server Action breaks the app router once a Parallel slot is rendered

### DISCUSSION THREAD

**Discussion by steve-marmalade** [vercel/next.js#63900] [Help]
### Summary

I could use help either:
- Building a reproduction
- Sharing ideas on what the root cause could be

@ztanner, this is somewhat of an FYI for you. You've been crushing the Github Issues regarding Parallel + Intercepting routes, and I've been trying to do my part in testing the latest canaries. In general the feature has been much more stable, but I have hit a road block in our closed-source repo that I've been struggling to build a reproduction for. Rather than wait until I can narrow it down, I figured I'd at least start a discussion in case you (or others) have suggestions.

Here's the sequence:
1. Load content in a Parallel Route (in my case, calling `router.replace('/path/that/matches/slot', {scroll: false})` to add new content to an existing page as a user scrolls down).
2. Trigger a server action (e.g. click a button)
3. In the server action, call `revalidateTag` or `revalidatePath`
4. Notice several issues with the router:
    1. The results of the revalidation are not displayed in the UI
        - I can see the `fetch-server-response.ts` events in the network tab with what looks like a valid RSC payload, but they aren't displayed 
    2. Subsequent navigations to _parallel/intercepted routes_ no longer work (I can see that the event was triggered and the server prepared the RSC, but the URL never changes and the slot is never loaded)
    3. Subsequent navigations to regular pages seem to work, but the app router is in a buggy/bad state: e.g. things like backward navigation don't work properly (the address bar updates but the UI doesn't).

**Comment by ztanner** [comment]
Hey @steve-marmalade ! Thanks for testing out the canaries. Will indeed be a bit tricky to troubleshoot without a reproduction. One thing I'm curious about though: if you swap out calling a server action + revalidate with instead calling `router.refresh`, is the reproduction the same? Refresh & server action handling trigger different codepaths so might be helpful to narrow that part down.

**Comment by steve-marmalade** [comment]
@ztanner , thanks for the quick response!

Yes, the reproduction is the same for `router.refresh`. I guess that's a win for consistency if nothing else :) I'll keep working at the reproduction, and please don't hesitate to throw out any other things to try.

Quick clarification on `router.refresh` vs `revalidateTag` if you have sec. My previous understanding of the benefits of `revalidateTag` was that:
1. There's only a single server round trip (the refreshed data comes back to the client as part of the mutation)
2. It enables more fine-grained cache updates (since it only needs to re-fetch the fetches/unstable_cache functions that have that specific cache tag).

However, based on https://github.com/vercel/next.js/issues/63509, am I correct that only the first one is actually a current benefit? I.e. _any_ revalidateTag will trigger a refetch for _all_ fetch requests on the page, and so we'd need to use a traditional client state management library like `useSWR` if we want to do more fine-grained revalidations?

**Reply by ztanner** [reply]
I could also look into a Replay of the issue (via https://replay.io/) if that's something you're willing to do. Is the data you're expecting to become visible after revalidation part of the parallel/interception route? Or is it data for something else that is also rendered on the page?

Not 100% sure on the revalidation question -- I'll need to take a closer look at the issue!

**Reply by steve-marmalade** [reply]
> Is the data you're expecting to become visible after revalidation part of the parallel/interception route? Or is it data for something else that is also rendered on the page?

I don't see refreshed data in either the content on the parallel route or in components elsewhere on the page. The same is true for clicking links that should navigate to a parallel/intercepted route (it doesn't matter if the link is in a component that's on the current parallel route, or elsewhere on the page).

I've recorded three replays to try and illustrate:

1. [Demonstrate expected behavior](https://app.replay.io/recording/replay-of-web-app-git-parallel-route-revalidate-marmaladevercelapp--ca6434ac-bffb-4a28-a2cf-a294a2889588)

- Everything in view is in the `page.tsx` of a dynamic route (`/q/[query]/page.tsx`)
- Toggling the upvote button triggers a server action and immediately updates the button UI via `useOptimistic`
- The total count of upvotes updates once the server action `revalidateTag` is processed
- The product feed also updates (the product card is no longer in the top row since it was downvoted)
- Clicking the Comment badge triggers a Modal implemented with Parallel + Intercepted routes, and this loads as expected
- Scrolling down the page triggers soft navigations to parallel routes `/q/[query]/2`, `/q/[query]/3`, etc and more product cards are loaded

2. [Demonstrate the issue after navigating to Parallel Route](https://app.replay.io/recording/replay-of-web-app-git-parallel-route-revalidate-marmaladevercelapp--0e0e197d-18db-4a17-846a-aed949535924)

- First we scroll down the page, so that `/q/[query]/2` is loaded
- We scroll back up and try to follow the exact same steps by toggling upvote.
- The optimistic update continues to work as expected
- But we don't see the upvote total update, and the product feed doesn't refresh as it did previously
- Clicking on the Comment badge no longer opens the Modal
- Scrolling down no longer successfully loads subsequent pages e.g. we see it attempt to navigate to `/q/[query]/3` but it never loads.

3. [Demonstrating the issue when interacting with the currently matched slot](https://app.replay.io/recording/replay-of-web-app-git-parallel-route-revalidate-marmaladevercelapp--8d408674-0840-40c5-b016-96f379fb57c8)

- Scroll down to product cards that were loaded as part of `/q/[query]/2`
- Stop before `/q/[query]/3` loads, so we are interacting with cards that are loaded in the current slot
- Toggle the upvote button
- The optimistic update works as usual
- Notice the product feed does not update to reflect the re-ordering
- Clicking on the Comment badge no longer opens the Modal

**Reply by ztanner** [reply]
Thank you! Totally spaced on mentioning this, but any chance of recording the second Replay with sourcemaps? I'll need to be able to step through Next.js' code execution to try and reverse engineer what's happening without access to the full source 🙏

**Reply by steve-marmalade** [reply]
No problem, let me know if this recording has the information you need! https://app.replay.io/recording/replay-of-localhost3000--815889c8-e802-4b51-a9c1-3e9236a90b73

**Reply by steve-marmalade** [reply]
Here's one that does `router.refresh()` instead of the optimistic update + server action + revalidate. I also removed some of our analytics libraries (sentry, segment, etc), in case that makes it easier to review.

https://app.replay.io/recording/19767070-b0e2-4483-ad72-6325baa166ed

**Reply by ztanner** [reply]
This is perfect, thank you! Will need to dig into it a bit more but I definitely see where it’s hanging.

**Reply by ztanner** [reply]
Good news! I've managed to create a reproduction, and it has to do with dynamic parameters (ie `[query]`). 

I have a fix that seems to work locally, will just need to write some tests tomorrow and I'll put up a PR! Thanks for helping track this down.

**Reply by steve-marmalade** [reply]
That's awesome! Thanks so much for digging into this.

**Reply by ztanner** [reply]
- https://github.com/vercel/next.js/pull/64006

if you'd like to keep an eye on the progress :)

**Reply by ztanner** [reply]
@steve-marmalade this should be published now in [v14.2.0-canary.55](https://github.com/vercel/next.js/releases/tag/v14.2.0-canary.55), mind giving that a try when you have a second and let me know if it resolves the issue(s)?

**Reply by steve-marmalade** [reply]
@ztanner - thank you so much for getting that out so quickly. Unfortunately, I am still seeing the same issue in the latest canary.

[Here's another replay](https://app.replay.io/recording/replay-of-localhost3000--3c6fdc56-7e42-428b-8ae0-2538131512ea), which just does `router.refresh()` when I click the upvote button. Just lmk if you want a replay for the Server Action + Revalidate code path :pray:

**Reply by ztanner** [reply]
Darn, ok, still looking into it! Question though... I noticed that there are these `?c=` query params that are getting injected into the URL via a `location.replace`. Is that coming from your code / is it something you could disable?

**Reply by steve-marmalade** [reply]
Fascinating, I think you are onto something. I am adding that search param as part of calling `router.replace` to navigate to the parallel route (e.g. `/q/housewarming` -> `/q/housewarming/2?c=abc123`). This gives the server some extra state for use in our business logic, but it's easy to remove for the sake of creating our reproduction. And ... tada! When I remove the `?c` search param, I can no longer reproduce the issue.

To try and isolate further, if I call `router.replace` with any query params (e.g. `?test=1`) the problem recurs (and it doesn't matter whether those search params are being referenced in the RSC). 

Concretely, `router.replace('/path/that/matches/parallel/route?any=param')` follow by a `router.refresh()` seems to trigger the issue. I can try to reproduce if that doesn't give you enough to go on.

**Reply by ztanner** [reply]
Ok nice, that does seem to be the issue here. Indeed it's fairly easy to reproduce once you add a query param -- I'll work on a fix + test case

**Reply by steve-marmalade** [reply]
I'm sure you're faster/better than I am at crafting test cases, but here's a demo in case it helps:

https://github.com/marmalade-labs/parallel-routes-revalidation-app-router
https://parallel-routes-revalidation-app-router.vercel.app/

1. Click on Upvote, see the numbers increment
2. Click on Load More
3. Click Upvote again, see the numbers increment
4. Click on "Load More (breaks)" (which just adds a random query param)
5. Click Upvote again, see that numbers don't increment
6. Click on either Load More link, the next page doesn't load

(I used vercel/kv so it was easier to visualize, but I expect `router.refresh` would have the same effect)

**Reply by steve-marmalade** [reply]
BTW - now that I have public repro, is it helpful for your tracking if I create a Github Issue? And does this one seem trickier to fix than the dynamic params issue / do you advise that I start implementing a workaround while you research?

**Reply by ztanner** [reply]
No worries! It should be resolved by https://github.com/vercel/next.js/pull/64086 :)

**Reply by steve-marmalade** [reply]
Incredible! Will follow along there.

Quick aside: would you say that it's safe to `redirect` from within a `default.tsx` if I don't want a parallel route to be navigable by itself? e.g. scrolling from `/q/[query]` to `/q/[query]/2` should be allowed, but navigating directly to `/q/[query/2` (e.g. via browser refresh, or clicking a link in slack) should redirect to `/q/[query]`.

My thought is to put the redirect in `/q/[query]/default.tsx`, which seems to work, although when I add a `console.log`, I do see that this `default.tsx` gets rendered sometimes even when non-default content is visible, like when I do back/forward navigation. When this happens, the redirect itself doesn't actually complete and I haven't observed any obvious negative consequences, but figured I'd ask in case you think there are drawbacks / have another suggestion for how to detect which slots have content.

My other idea was to check for `next-url` header in `middleware.ts` to detect a "hard navigation" and then redirect if the URL matches a regex for e.g. `q/[query]/2`, but I've seen elsewhere that applications are discouraged from depending on `next-url` directly.

**Reply by ztanner** [reply]
Redirecting in `default` seems reasonable to me! I'm curious about the `console.log` thing you're describing but if it's not actually being rendered, then it won't trigger our `RedirectBoundary` which handles the redirect.

**Reply by steve-marmalade** [reply]
That's helpful info thanks! Will see if I can whip up a simple repro for the `console.log` thing.

The new [v14.2.0-canary.58](https://github.com/vercel/next.js/releases/tag/v14.2.0-canary.58) seems to be working well for me so far! Will keep you posted.
